charts/mimir-distributed/k8s-diff/rules/services.yaml

```diff
@@ -0,0 +1,94 @@
+---
+patch_rules:
+
+- name: "Rename http ports in Services"
+  match:
+  - op: test
+    path: /kind
+    value: Service
+  - op: test
+    path: /spec/ports/0/port
+    value: 8080
+  steps:
+  - op: replace
+    path: /spec/ports/0/name
+    value: http-metrics
+  - op: replace
+    path: /spec/ports/0/targetPort
+    value: http-metrics
+
+- name: "Rename grpc ports in Services"
+  match:
+  - op: test
+    path: /kind
+    value: Service
+  - op: test
+    path: /spec/ports/1/port
+    value: 9095
+  steps:
+  - op: replace
+    path: /spec/ports/1/name
+    value: grpc
+  - op: replace
+    path: /spec/ports/1/targetPort
+    value: grpc
+
+- name: "Remove overrides exporter grpc port"
+  match:
+  - op: test
+    path: /kind
+    value: Service
+  - op: test
+    path: /metadata/name
+    value: overrides-exporter
+  - op: test
+    path: /spec/ports/1/name
+    value: grpc
+  steps:
+  - op: remove
+    path: /spec/ports/1
```

Comment:
Using indexes is slightly flaky. I have't read the code, but will this fail if the grpc port is the first one not the second one?

The docs say that the validation stage of yaml-patch should do that, but I'm not sure I understand those correctly. I see some of the patch rules explicitly work only on fields that exist in helm, so they can't affect jsonnet too. Yet, obviously, the yaml-patch succeeds for them.

Reply:
You're 100% right using an array index is flaky. Unfortunately jsonpointer doesn't have any smarter way to identify an array element. I've considered extending it, but have avoided that so far in order to hopefully keep the number of net-new concepts in this tool to a minimum.

In the case of this rule, if someone changed the grpc port to come first, then the `match` would not succeed. (`/spec/ports/1/name` != `grpc`). Then during validation `yaml-patch` will report that this rule didn't match any objects and will exit 1. It's not ideal, since the error message doesn't necessarily point out the actual problem. 

Secondly, each rule only has to effect a single object from either side of the diff in order to be considered valid. There are quite a few rules that only impact helm for example.